keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1709 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this enhancement, Kubernetes Developers will declare validation rules using
+IDL tags in the `types.go` files that to define the Kubernetes native API types.
+For example:
+
+```go
+// staging/src/k8s.io/api/core/v1/types.go
+
+// +validationRule="!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
+type PodSpec struct {
+  Containers []Container `json:...`
+}
+
+type Container struct {
+  // ...
+
+  Ports []ContainerPort `json:...`
+
+  // ...
+}
+
+type ContainerPort struct {
+  // ...
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  HostPort int32 `json:...`
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  ContainerPort int32 `json:...`
+
+  // ...
+}
+```
+
+In this example, both `+validation`, `+minimum` and `+maximum` are IDL tags.
+
+The declarative validation rules will be used by the kube-apiserver to
+validate API requests.
+
+The declarative validation rules will also be included in the published OpenAPI:
+
+```json
+// Pod Spec
+"openAPIV3Schema": {
+  "type": "object",
+  "x-kubernetes-validations": [
+    {
+      "rule": "!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
+    }
+  ],
+}
+
+...
+
+// Container Port
+"openAPIV3Schema": {
+  "type": "object",
+  "properties": {
+    "hostPort": {
+      "minimum": 1,
+      "maximum": 65535
+      ...
+    }
+    "containerPort": {
+      "minimum": 1,
+      "maximum": 65535
+      ...
+    }
+  }
+}
+```
+
+It is important to note that the declarative validation will use the same set of
+validation options available today for CRDs. namely:
+
+- [JSON Schema value validations](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00) (e.g. `format`, `maxItems`)
+- [CEL Validation Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) (e.g. `self.replicas <= self.maxReplicas`)
+- Relevant [merge strategy declarations](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) (e.g. `listType=set`)
+
+## Motivation
+
+Kubernetes API validation rules are currently written by hand, which makes them
+difficult for users to access directly.
+
+Declarative validation will benefit Kubernetes maintainers:
+
+- It will make it easier to develop, maintain and review APIs.
```

Comment:
I agree putting all of the validation, error, reason, etc. in one line is not good UX for writing or reviewing, which is counter to our goal.

I think at the very least we should have the ability to split the annotation for the rule, reason, fieldPath into separate lines. We can add this to the code generator. I will add that to our list of requirements for adding x-kubernetes-validations marker comments to the code generator.

Another QoL improvement which would assist in expression writing might be to support variables as we do for ValidatingAdmissionPolicy. I think that is out of scope for this KEP since it would need to be added to CRD Validation Rules (KEP will just use the same API exposed)

I also think referring to constants from code is something to be explored, but seems challenging

Reply:
What if we had a tool to help authoring/testing/maintaining these?

Reply:
Tools help, but ultimately humans have to read and review them.

Reply:
I second Rob's point - I saw the cel logic in the Gateway API and it is very difficult to read. I think the mechanic of using comments is very brittle and doesn't enable having a tight feedback loop (eg. running unit tests)

Secondly, if you have the same constant in go code and cel logic (for whatever reason) it's extra taxing to keep them in sync.

To offer some solutions:
1. Move the cel logic to their own files and allow them to be read in unit tests (kubebuilder or other tooling can probably automate this). We'd need a way to expose some go constants to this cel env.
2. Allow validation to be written in go and transform the AST to cel using codegen (probably intractable)

Reply:
@dprotaso there's a balance to strike here - a big part of this (for me) is making it so that I don't have to bounce between files when reviewing most APIs. That suggests in-line validation >> out-of-line.

I think we should explore multi-line tags and loading constants (as we can with +default)

Reply:
Folks, it's hard for us to reason without looking at the more intricate details of what makes these expression hard to read/write. Do we have some examples we can refer to so that we can have a better understanding of what's going on?

Reply:
@dprotaso @thockin Ideally if there are constants in use, you could publish them as part of the CEL environment without too much trouble, though you may want to namespace them, e.g. `k8s.<ConstantName>`

@alexzielenski I'm not sure what the appetite for this is, but CEL currently has support for exhaustive evaluation and tracing, so it should be possible to track the expression evaluation behavior back to the point in the file where it's defined and highlight evaluation state, sort of like a poor version of code coverage, but highly visual in nature. This would probably need to be its own tool. CEL team is also interested in generating lcov-style coverage information for a given input which may be useful for such dev/debug flows.

Reply:
Just to be clear, I'm very glad we have the option to use CEL in CRDs, it's a huge improvement over validating webhooks. With that said, as @apelisse suggested, it might be helpful to look at an example of what the experience is like for us today. Let's walk through a recent change I made to our CEL validation in Gateway API: https://github.com/kubernetes-sigs/gateway-api/pull/2433.

We have a list of `ParentRefs` that we each want to be distinct. The original validation required that when Group, Kind, Name, and Namespace we're identical, one of Port or SectionName also needed to be set and different. This meant that one ParentRef could set Port and the other could set SectionName and those ParentRefs would be considered unique. The validation to initially implement that was done in https://github.com/kubernetes-sigs/gateway-api/pull/2320. As far as we could tell, we actually needed two fairly complex CEL validations to accomplish this: 

1. sectionName or port must be specified when parentRefs includes 2 or more references to the same parent
2. sectionName or port must be unique when parentRefs includes 2 or more references to the same parent

There's a more detailed overview of that logic that was used in the PR description. To further complicate matters, we needed to write 2 versions of both of these validations because our Standard release channel does not include the Port field, and our Experimental release channel does. (That's not CEL's fault, but with upstream Go validation I think we'd be able to just have an inline check for feature gate enablement for example).

Then it comes to the actual workflow. Here's what my experience was like:

1. Try to understand the current validation by manually splitting it out from it's condensed single-line format into a multi-line format that would be easier to read. 
2. We haven't found a way to include comments explaining the logic of individual CEL lines/checks, but fortunately the PR that added the validation had a good description, so I referred to that and talked with the author of the original validation to make sure I fully understood everything. 
3. I then fumbled around the [CEL Playground](https://playcel.undistro.io/?content=H4sIAAAAAAAAA7VRTWuEMBD9K9McjIIVzFHYv9DrXhZCWKdWqjEYSy3d%2FvcmMX5uYHtpTjNv5r15vHyTUgyCFERj81pcJJj3DFXffagCKI6iVQ3mWddXdBoCvNeyNLNzXVY4LKgULRq0%2Ffrc45zbiVbiipybhQH1OtR4HepOvkxc3zH6DzaCl%2FJlTFKCo%2BpRawP7NDLRNLHKU3ANjrUeNO8kxoqlYIkqz5xDOJ1AMV9HkcWtOQ%2B7ckKtOY%2B60qBWJ55dLIVrnt6ENvezXYIJ3G5wBK0mpcmsd1BgIQUWVJjZbifgKuzInA0fiqKVGvJ8ZCznbZH8MZ7Nz87hbKBH0dyxWYD9KJYlmJ3aGssB3hLv3e7395FcZJKQn184gYhztAMAAA%3D%3D) trying to make this work. This is a great tool that I really appreciate, but unfortunately still has some rough edges like not being able to resize portions + losing work if you accidentally navigate away.
4. Once I had something that I thought would work, I then manually condensed it back into a single line statement and added it to our CRD annotations.
5. I then updated our go tests that cover this validation. To run them, we have to generate updated versions of our CRDs, deploy them to a cluster, and then run our tests against that cluster. I'm hoping there's a better workflow available here, but in practice it means that each change to the CRD validation has a fairly long feedback loop.

All of the above took me far longer than I'd like to admit, and like 10x longer than it would've taken me to write and test an equivalent validation with go. I'm hoping that there are ways we can improve our process so it's not quite as painful for other contributors in the future.

Reply:
Thanks for the specifics, Rob.  As we collectivel learn how to do this, you are again paying the trail-blazer tax :)

A few things to think about:

1) We don't need to block the KEP on *solving* this, just acknowledging that it needs to be solved.

2) There are going to be "fairly trivial" validations that are easy to read and comprehend.  There are going to be "fairly complex" validations that would benefit from being written as multi-line expressions (maybe with comments!) which seems like a k8s problem rather than intrinsic to CEL.  And there will be some "very complex" validations which do not easily distill into a single expression.  For this last set, we need to consider those specific cases.  It may be that there are idioms or capabilities missing in CEL that we can help define.  It may be that there are library functions missing.  It may be that some validations just DON'T fit very well.

3) Our near-to-medium term goal is not 100% elimination of hand-written validation.  Let's take wins where we can get them and iterate on how to fit the hard-cases.

4) I want to believe that it's in everyone's best interest to make CEL work well here, so I will assume that reasonable questions will be met with reasonable consideration until otherwise demonstrated :)  Let's not assume that everything is fully baked and we have no degrees of freedom.

Reply:
> https://github.com/kubernetes-sigs/gateway-api/pull/2433.

![image](https://github.com/kubernetes/enhancements/assets/1929644/6ba3c42a-c4d3-4f28-aabb-199a97b17838)
I suggest we create "The International Obfuscated CEL Code Contest" (see https://www.ioccc.org/ for those not familiar with it). 🤣

Reply:
Completely agree with everything @thockin said, not trying to block this KEP at all, just trying to find ways we can improve the overall experience here.

Reply:
@robscott Thanks for sharing the detailed notes about your experience. It's useful in how I consider which sorts of tools the CEL team should consider building in the future.

Reply:
> This is critical feedback. I agree that this KEP should be blocked from GA, maybe even from beta, by positive feedback from some preponderance of use-cases (e.g. converting "core" validation to use it).

I share reservations about whether CEL is actually easier than golang, particular for feature-gated fields, tests that require oldSelf, and cross field references.  I agree this does not block alpha.  I think it should prevent beta (on by default).